core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -2580,6 +2581,90 @@ private void loadMetadataIfNotExist(RpcContext rpcContext, LockedInodePath inode
     }
   }
 
+  @Override
+  public void updateMount(AlluxioURI alluxioPath, MountContext context)
+      throws FileAlreadyExistsException, FileDoesNotExistException, InvalidPathException,
+      IOException, AccessControlException {
+    LockingScheme lockingScheme = createLockingScheme(alluxioPath,
+        context.getOptions().getCommonOptions(), LockPattern.WRITE_EDGE);
+    try (RpcContext rpcContext = createRpcContext();
+         LockedInodePath inodePath = mInodeTree
+             .lockInodePath(lockingScheme.getPath(), lockingScheme.getPattern());
+         FileSystemMasterAuditContext auditContext = createAuditContext(
+             "updateMount", alluxioPath, null, inodePath.getParentInodeOrNull())) {
+      try {
+        mPermissionChecker.checkParentPermission(Mode.Bits.WRITE, inodePath);
+      } catch (AccessControlException e) {
+        auditContext.setAllowed(false);
+        throw e;
+      }
+      MountInfo mountInfo = mMountTable.getMountTable().get(alluxioPath.getPath());
+      if (mountInfo == null) {
+        throw new InvalidPathException("Failed to update mount properties for "
+            + inodePath.getUri() + ". Please ensure the path is an existing mount point.");
+      }
+      List<AlluxioURI> syncPoints = null;
+      try (LockResource r = new LockResource(mSyncManager.getSyncManagerLock())) {
+        syncPoints = mSyncManager.getFilterList(mountInfo.getMountId());
+        // test the new mount configuration first before unmounting the old ufs
+        UnderFileSystemConfiguration ufsConf = UnderFileSystemConfiguration
+            .defaults(ServerConfiguration.global())
+            .createMountSpecificConf(context.getOptions().getPropertiesMap());
+        String ufsUri = mountInfo.getUfsUri().toString();
+        try (UnderFileSystem ufs = UnderFileSystem.Factory.create(ufsUri, ufsConf)) {
+          // check the root path is valid
+          ufs.getFileStatus(ufsUri);
+          // check the sync points are valid
+          if (syncPoints != null && !syncPoints.isEmpty()) {
+            for (AlluxioURI syncPoint : syncPoints) {
+              ufs.startSync(syncPoint);
+            }
+          }
+        } catch (Throwable t) {
+          throw new InvalidArgumentException("Failed to mount under file system."
+              + " Please ensure the options are valid", t);
+        }
+        unmountInternal(rpcContext, inodePath, false);
+      }
+      long mountId = IdUtils.createMountId();
+      try {
+        mountInternal(rpcContext, inodePath, mountInfo.getUfsUri(), mountId, context, true);
```

Comment:
should there be a `updateMountInternal` and an associated `updateMount` in the mount table? It looks like we are reusing delete and add, but they are not performed atomically, which could cause strange behavior?

Reply:
I want to reuse the `unmountInternal` and `mountInternal` to minimize the risk of having a new code path and get out of sync in the future when we add new logic to only one of the function. 

Regarding atomicity, the current implementation use a single write lock on the mount point and also does a dry run to load the UFS client with new configuration options before making change to the mount table, so it should be safe to use 99% of the time. If we think it is crucial to achieve atomicity for the rest 1% of the cases, I can refactor the mount table to support it but I feel that will also make mount table less maintainable.

Reply:
My concern would not be that other threads have write access to the inode, but rather, the update mount operation itself. It is represented by 2 journal entries in sequence: unmount and mount. If something happens in between, then the resulting state would be very inconsistent?